include/range/v3/algorithm/adjacent_remove_if.hpp

```diff
@@ -0,0 +1,89 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler
+//  Copyright Christopher Di Bella
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_ALGORITHM_ADJACENT_REMOVE_IF_HPP
+#define RANGES_V3_ALGORITHM_ADJACENT_REMOVE_IF_HPP
+
+#include <range/v3/algorithm/adjacent_find.hpp>
+#include <range/v3/algorithm/move.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/utility/functional.hpp>
+#include <range/v3/utility/iterator_concepts.hpp>
+#include <range/v3/utility/iterator_traits.hpp>
+#include <range/v3/utility/static_const.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-algorithms
+        /// @{
+        struct adjacent_remove_if_fn
+        {
+            /// \brief function \c adjacent_remove_if_fn::operator()
+            ///
+            /// range-based version of the \c adjacent_remove_if algorithm
+            ///
+            /// \pre `R` is a model of the `ForwardRange` concept.
+            /// \pre `Pred` is a model of the `BinaryPredicate` concept.
+            template<typename I, typename S, typename Pred, typename Proj = ident,
+                CONCEPT_REQUIRES_(Permutable<I>() && Sentinel<S, I>() &&
+                    IndirectRelation<Pred, projected<I, Proj>>())>
+            I
+            operator()(I first, S last, Pred pred = {}, Proj proj = {}) const
+            {
+                first = adjacent_find(std::move(first), last, std::ref(pred), std::ref(proj));
+                if (first != last)
+                {
+                    auto i = next(first);
+                    if (i != last)
```

Comment:
`i != last` is always true here:
The postcondition of `adjacent_find` is that it either returns `last` or the first iterator `x` such that `pred(proj(x), proj(x+1))` is `true`. If `first != last`, then the predicate must have held, so `next(first)` must be a dereferenceable iterator and `next(first) != last` must be true.